File: Pokemon-Objc/Pokemon-Objc/PokemonDetailViewController.m
//
//  PokemonDetailViewController.m
//  Pokemon-Objc
//
//  Created by Chad Parker on 8/14/20.
//  Copyright © 2020 Chad Parker. All rights reserved.
//

#import "PokemonDetailViewController.h"
#import "Pokemon_Objc-Swift.h"
#import "LSIPokemon.h"

void *KVOContext = &KVOContext;

@interface PokemonDetailViewController ()

@property (strong, nonatomic) IBOutlet UIImageView *imageView;
@property (strong, nonatomic) IBOutlet UILabel *nameLabel;
@property (strong, nonatomic) IBOutlet UILabel *abilitiesListLabel;
@property (strong, nonatomic) IBOutlet UILabel *idNumberLabel;

@end

@implementation PokemonDetailViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self clearLabels];
    
    if (self.pokemon) {
        [self updateViews];
        [PokemonAPI.sharedController fillInDetailsFor:self.pokemon];
        [self.pokemon addObserver:self forKeyPath:@"identifier" options:0 context:KVOContext];
    }
}

- (void)clearLabels
{
    self.nameLabel.text = nil;
    self.abilitiesListLabel.text = nil;
    self.idNumberLabel.text = nil;
}

- (void)updateViews
{
    self.nameLabel.text = [self.pokemon.name capitalizedString] ?: @"";
    self.idNumberLabel.text = [NSString stringWithFormat:@"%@", self.pokemon.identifier] ?: @"";
    self.abilitiesListLabel.text = [self.pokemon.abilities componentsJoinedByString:@"\n"];
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSKeyValueChangeKey,id> *)change context:(void *)context
{
    if (context == KVOContext) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [self updateViews];
            [self fetchAndUpdateImage];
        });
    } else {
        [super observeValueForKeyPath:keyPath ofObject:object change:change context:context];
    }
}

- (void)fetchAndUpdateImage
{
    [PokemonAPI.sharedController fetchImageAt:self.pokemon.spriteURLString completion:^(UIImage * _Nullable image) {
        dispatch_async(dispatch_get_main_queue(), ^{
            self.imageView.image = image;
        });
    }];
}

- (void)dealloc
{
    [self.pokemon removeObserver:self forKeyPath:@"identifier" context:KVOContext];
}

@end

File: Pokedex 2.0/Pokedex 2.0/View Controllers/PokemonDetailViewController.swift
//
//  PokemonDetailViewController.swift
//  Pokedex 2.0
//
//  Created by Moses Robinson on 4/5/19.
//  Copyright © 2019 Moses Robinson. All rights reserved.
//

import UIKit

class PokemonDetailViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        
        PokemonAPI.addObserver(self, forKeyPath: "pokemon", options: [], context: &KVOContext)
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        fetchPokemon()
    }
    
    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
        
        if context == &KVOContext {

            DispatchQueue.main.async {
                self.updateViews()
            }

        } else {
            super.observeValue(forKeyPath: keyPath, of: object, change: change, context: context)
        }
    }
    
    private func updateViews() {
        
        guard let pokemon = PokemonAPI.shared.pokemon else { return }
        
        nameLabel.text = "Name: \(pokemon.name.capitalized)"
        idLabel.text = "ID: \(pokemon.identifier!)"
        
        let abilitiesString = pokemon.abilities.map {$0}?.joined(separator: ", ").capitalized
        abilitiesLabel.text = "Abilities: \(abilitiesString!)"
        
        
    }
    
    private func fetchPokemon() {
     
        guard let pokemon = self.pokemon else { return }
        
        PokemonAPI.shared.fillInDetails(for: pokemon)
    }
    

    // MARK: - Properties
    
    @objc var pokemon: Pokemon?
    private var KVOContext = 0
    
    @IBOutlet var pokemonSprite: UIImageView!
    @IBOutlet var nameLabel: UILabel!
    @IBOutlet var idLabel: UILabel!
    @IBOutlet var abilitiesLabel: UILabel!
}
